File: src/lib/backtest.ts
import type {
  StrategyDefinition,
  Condition,
  ConditionOperator,
  OHLCVBar,
  BacktestResult,
  TradeOrder,
  OpenPosition,
  ClosedTrade,
  EquityPoint,
} from "./types"
import { computeIndicator, getIndicatorValue, type IndicatorResult } from "./indicators"

function cryptoId(): string {
  const arr = new Uint8Array(16)
  for (let i = 0; i < 16; i++) arr[i] = Math.floor(Math.random() * 256)
  return Array.from(arr, b => b.toString(16).padStart(2, "0")).join("")
}

function defaultStrategy(s: StrategyDefinition): Required<Pick<StrategyDefinition, "timeframe" | "initial_capital" | "exit_rules" | "risk_management" | "position_sizing" | "transaction_costs">> & StrategyDefinition {
  return {
    ...s,
    timeframe: s.timeframe || "1d",
    initial_capital: s.initial_capital || 100000,
    exit_rules: s.exit_rules && s.exit_rules.length > 0 ? s.exit_rules : [{}],
    risk_management: s.risk_management || {},
    position_sizing: s.position_sizing || {},
    transaction_costs: s.transaction_costs || {},
  }
}

function evaluateCondition(cond: Condition, indicatorCache: Record<string, IndicatorResult>, bars: OHLCVBar[], idx: number): boolean {
  const period = cond.period || 14
  const cacheKey = `${cond.indicator}_${period}`
  if (!indicatorCache[cacheKey]) {
    indicatorCache[cacheKey] = computeIndicator(cond.indicator, bars, { period })
  }
  const actual = getIndicatorValue(indicatorCache[cacheKey], cond.indicator, idx)
  const target = typeof cond.value === "number" ? cond.value : parseFloat(String(cond.value)) || 0

  switch (cond.operator) {
    case ">": return actual > target
    case "<": return actual < target
    case ">=": return actual >= target
    case "<=": return actual <= target
    case "==": return Math.abs(actual - target) < 1e-8
    case "!=": return Math.abs(actual - target) >= 1e-8
    case "cross_above": {
      if (idx < 1) return false
      const prev = getIndicatorValue(indicatorCache[cacheKey], cond.indicator, idx - 1)
      return prev <= target && actual > target
    }
    case "cross_below": {
      if (idx < 1) return false
      const prev = getIndicatorValue(indicatorCache[cacheKey], cond.indicator, idx - 1)
      return prev >= target && actual < target
    }
    default: return false
  }
}

function evaluateConditions(conditions: Condition[], indicatorCache: Record<string, IndicatorResult>, bars: OHLCVBar[], idx: number): boolean {
  return conditions.every(c => evaluateCondition(c, indicatorCache, bars, idx))
}

function calculateCharges(price: number, quantity: number, side: string, tc: StrategyDefinition["transaction_costs"] = {}) {
  const turnover = price * quantity
  const brokerage = tc?.brokerage_per_order ?? 20
  const sttPct = tc?.stt_percent ?? 0.05
  const exchangeCharges = tc?.exchange_charges ?? 0.003
  const sebiCharges = tc?.sebi_charges ?? 0.001
  const gstPct = tc?.gst_percent ?? 18
  const slippageTicks = tc?.slippage_ticks ?? 1
  const slippageValue = tc?.slippage_value ?? 0

  const stt = turnover * (sttPct / 100)
  const exchange = turnover * (exchangeCharges / 100)
  const sebi = turnover * (sebiCharges / 100)
  const gst = (brokerage + exchange) * (gstPct / 100)
  const slippage = slippageTicks * slippageValue * quantity

  return { brokerage, stt, other_charges: exchange + sebi + gst, slippage }
}

function fillOrder(
  timestamp: string, instrument: string, side: string,
  quantity: number, price: number, tag: string,
  capital: number, tc: StrategyDefinition["transaction_costs"] = {}
): { order: TradeOrder; capital: number } {
  const slippageTicks = tc?.slippage_ticks ?? 1
  const slippageValue = tc?.slippage_value ?? 0
  const slippage = slippageTicks * slippageValue
  const fillPrice = side === "buy" ? price + slippage : price - slippage
  const charges = calculateCharges(fillPrice, quantity, side, tc)
  const totalCost = charges.brokerage + charges.stt + charges.other_charges

  if (side === "buy") capital -= fillPrice * quantity + totalCost
  else capital += fillPrice * quantity - totalCost

  return {
    order: {
      id: cryptoId(),
      timestamp,
      instrument,
      side,
      quantity,
      price: fillPrice,
      slippage: charges.slippage,
      brokerage: charges.brokerage,
      stt: charges.stt,
      other_charges: charges.other_charges,
      tag,
    },
    capital,
  }
}

function computePositionPnl(pos: OpenPosition, currentPrice: number): number {
  const entryPrice = pos.entry_order.price
  const qty = pos.entry_order.quantity
  return pos.entry_order.side === "buy"
    ? (currentPrice - entryPrice) * qty
    : (entryPrice - currentPrice) * qty
}

export function runBacktest(strategy: StrategyDefinition, bars: OHLCVBar[]): BacktestResult {
  const s = defaultStrategy(strategy)
  const rm = s.risk_management
  const maxPositions = rm.max_positions ?? 1
  const tc = s.transaction_costs
  const ps = s.position_sizing
  const initialCapital = s.initial_capital

  let capital = initialCapital
  let peakEquity = initialCapital
  const positions: OpenPosition[] = []
  const closedTrades: ClosedTrade[] = []
  const equityCurve: EquityPoint[] = []
  let totalBrokerage = 0
  let totalStt = 0
  const indicatorCache: Record<string, IndicatorResult> = {}

  for (let idx = 0; idx < bars.length; idx++) {
    const ts = bars[idx].datetime
    const currentPrice = bars[idx].Close

    const closedIndices: number[] = []
    for (let i = 0; i < positions.length; i++) {
      const pos = positions[i]
      if (currentPrice > pos.highest_price) pos.highest_price = currentPrice
      pos.bars_held += 1
      pos.current_price = currentPrice
      pos.unrealized_pnl = computePositionPnl(pos, currentPrice)

      let exitReason: string | null = null

      if (rm.stop_loss_percent && pos.entry_order.side === "buy") {
        const slPrice = pos.entry_order.price * (1 - rm.stop_loss_percent / 100)
        if (currentPrice <= slPrice) exitReason = "stop_loss"
      } else if (rm.stop_loss_percent && pos.entry_order.side === "sell") {
        const slPrice = pos.entry_order.price * (1 + rm.stop_loss_percent / 100)
        if (currentPrice >= slPrice) exitReason = "stop_loss"
      }

      if (!exitReason && rm.take_profit_percent && pos.entry_order.side === "buy") {
        const tpPrice = pos.entry_order.price * (1 + rm.take_profit_percent / 100)
        if (currentPrice >= tpPrice) exitReason = "take_profit"
      } else if (!exitReason && rm.take_profit_percent && pos.entry_order.side === "sell") {
        const tpPrice = pos.entry_order.price * (1 - rm.take_profit_percent / 100)
        if (currentPrice <= tpPrice) exitReason = "take_profit"
      }

      if (!exitReason && rm.trailing_stop_percent) {
        if (pos.entry_order.side === "buy") {
          const trailPrice = pos.highest_price * (1 - rm.trailing_stop_percent / 100)
          if (currentPrice <= trailPrice) exitReason = "trailing_stop"
        } else {
          const trailPrice = pos.highest_price * (1 + rm.trailing_stop_percent / 100)
          if (currentPrice >= trailPrice) exitReason = "trailing_stop"
        }
      }

      if (!exitReason) {
        for (const rule of s.exit_rules) {
          if (rule.max_bars_in_trade && pos.bars_held >= rule.max_bars_in_trade) { exitReason = "max_bars"; break }
          if (rule.exit_at_end_of_day) {
            const hour = parseInt(ts.split("T")[1]?.split(":")[0] || "0", 10)
            const minute = parseInt(ts.split("T")[1]?.split(":")[1] || "0", 10)
            if (hour >= 15 && minute >= 15) { exitReason = "eod"; break }
          }
          if (rule.conditions && rule.conditions.length > 0 && evaluateConditions(rule.conditions, indicatorCache, bars, idx)) {
            exitReason = "signal"; break
          }
        }
      }

      if (exitReason) {
        const exitSide = pos.entry_order.side === "buy" ? "sell" : "buy"
        const { order: exitOrder, capital: newCap } = fillOrder(ts, pos.entry_order.instrument, exitSide, pos.entry_order.quantity, currentPrice, exitReason, capital, tc)
        capital = newCap
        const netPnl = pos.unrealized_pnl - exitOrder.brokerage - exitOrder.stt - exitOrder.other_charges
        closedTrades.push({
          entry_order: pos.entry_order,
          exit_order: exitOrder,
          pnl: pos.unrealized_pnl,
          net_pnl: netPnl,
          holding_bars: pos.bars_held,
          exit_reason: exitReason,
        })
        capital += pos.unrealized_pnl
        totalBrokerage += exitOrder.brokerage
        totalStt += exitOrder.stt
        closedIndices.push(i)
      }
    }

    for (let i = closedIndices.length - 1; i >= 0; i--) {
      positions.splice(closedIndices[i], 1)
    }

    let dailyLoss = 0
    for (const trade of closedTrades) {
      const tradeDate = trade.exit_order.timestamp.split("T")[0]
      const currentDate = ts.split("T")[0]
      if (tradeDate === currentDate && trade.net_pnl < 0) {
        dailyLoss += Math.abs(trade.net_pnl)
      }
    }
    if (rm.max_loss_per_day && dailyLoss >= rm.max_loss_per_day) {
      equityCurve.push({ timestamp: ts, equity: capital, drawdown: 0 })
      continue
    }

    if (positions.length < maxPositions) {
      for (const rule of s.entry_rules) {
        if (evaluateConditions(rule.conditions, indicatorCache, bars, idx)) {
          const qty = getQuantity(ps, bars, capital)
          const { order, capital: newCap } = fillOrder(ts, rule.instrument, rule.side, qty, currentPrice, "entry", capital, tc)
          capital = newCap
          positions.push({
            entry_order: order,
            current_price: currentPrice,
            highest_price: currentPrice,
            unrealized_pnl: 0,
            bars_held: 0,
          })
          totalBrokerage += order.brokerage
          totalStt += order.stt
          break
        }
      }
    }

    if (capital > peakEquity) peakEquity = capital
    const dd = peakEquity > 0 ? ((peakEquity - capital) / peakEquity) * 100 : 0
    equityCurve.push({ timestamp: ts, equity: capital, drawdown: dd })
  }

  for (const pos of [...positions]) {
    const lastPrice = bars[bars.length - 1].Close
    const exitSide = pos.entry_order.side === "buy" ? "sell" : "buy"
    const { order: exitOrder, capital: newCap } = fillOrder(
      bars[bars.length - 1].datetime, pos.entry_order.instrument, exitSide,
      pos.entry_order.quantity, lastPrice, "end_of_backtest", capital, tc
    )
    capital = newCap
    const pnl = computePositionPnl(pos, lastPrice)
    closedTrades.push({
      entry_order: pos.entry_order,
      exit_order: exitOrder,
      pnl,
      net_pnl: pnl - exitOrder.brokerage - exitOrder.stt,
      holding_bars: pos.bars_held,
      exit_reason: "end_of_backtest",
    })
  }
  positions.length = 0

  return computeResults(s, bars, closedTrades, equityCurve, capital, peakEquity, totalBrokerage, totalStt, initialCapital)
}

function getQuantity(ps: StrategyDefinition["position_sizing"] = {}, bars: OHLCVBar[], equity: number): number {
  const sizingType = ps?.sizing_type || "fixed_lot"
  const close = bars[0].Close
  switch (sizingType) {
    case "fixed_lot": return ps?.lot_size || 1
    case "fixed_capital": return Math.max(1, Math.floor((ps?.capital_per_trade || 100000) / close))
    case "percent_capital": {
      const alloc = equity * (ps?.percent_capital || 10) / 100
      return Math.max(1, Math.floor(alloc / close))
    }
    default: return 1
  }
}

function computeResults(
  s: Required<Pick<StrategyDefinition, "initial_capital">> & StrategyDefinition,
  bars: OHLCVBar[],
  trades: ClosedTrade[],
  equityCurve: EquityPoint[],
  finalCapital: number,
  peakEquity: number,
  totalBrokerage: number,
  totalStt: number,
  initialCapital: number,
): BacktestResult {
  const n = trades.length
  const wins = trades.filter(t => t.net_pnl > 0)
  const losses = trades.filter(t => t.net_pnl <= 0)

  const totalDays = (new Date(bars[bars.length - 1].datetime).getTime() - new Date(bars[0].datetime).getTime()) / (1000 * 60 * 60 * 24)
  const years = Math.max(totalDays / 365.25, 0.01)
  const totalReturn = ((finalCapital - initialCapital) / initialCapital) * 100
  const cagr = (Math.pow(finalCapital / initialCapital, 1 / years) - 1) * 100

  const eqSeries = equityCurve.map(e => e.equity)
  const ddSeries = equityCurve.map(e => e.drawdown)
  const maxDD = ddSeries.length > 0 ? Math.max(...ddSeries) : 0

  const dailyReturns: number[] = []
  for (let i = 1; i < eqSeries.length; i++) {
    if (eqSeries[i - 1] > 0) dailyReturns.push((eqSeries[i] - eqSeries[i - 1]) / eqSeries[i - 1])
  }

  const avgRet = dailyReturns.length > 0 ? dailyReturns.reduce((a, b) => a + b, 0) / dailyReturns.length : 0
  const variance = dailyReturns.length > 0 ? dailyReturns.reduce((a, b) => a + (b - avgRet) ** 2, 0) / dailyReturns.length : 0
  const stdDev = Math.sqrt(variance)
  const volatility = stdDev * Math.sqrt(252) * 100

  const downside = dailyReturns.filter(r => r < 0)
  const downVariance = downside.length > 0 ? downside.reduce((a, b) => a + b ** 2, 0) / downside.length : 1e-20
  const downsideVol = Math.sqrt(downVariance) * Math.sqrt(252)
  const sharpe = (avgRet * 252) / (stdDev * Math.sqrt(252) || 1e-10)
  const sortino = (avgRet * 252) / (downsideVol || 1e-10)

  const avgWin = wins.length > 0 ? wins.reduce((a, t) => a + t.net_pnl, 0) / wins.length : 0
  const avgLoss = losses.length > 0 ? losses.reduce((a, t) => a + t.net_pnl, 0) / losses.length : 0
  const grossProfit = wins.reduce((a, t) => a + t.net_pnl, 0)
  const grossLoss = Math.abs(losses.reduce((a, t) => a + t.net_pnl, 0))
  const profitFactor = grossLoss > 0 ? grossProfit / grossLoss : 0
  const winRate = n > 0 ? (wins.length / n) * 100 : 0
  const avgBars = n > 0 ? trades.reduce((a, t) => a + t.holding_bars, 0) / n : 0

  const tradingDays = eqSeries.length
  const daysInPosition = trades.reduce((a, t) => a + t.holding_bars, 0)
  const exposure = tradingDays > 0 ? (daysInPosition / tradingDays) * 100 : 0

  const monthly: Record<string, { equity: number }> = {}
  for (const e of equityCurve) {
    const key = e.timestamp.substring(0, 7)
    monthly[key] = { equity: e.equity }
  }
  const monthlyReturns: { month: string; return: number; equity: number }[] = []
  let prevEq = initialCapital
  for (const [month, { equity }] of Object.entries(monthly).sort((a, b) => a[0].localeCompare(b[0]))) {
    const ret = prevEq > 0 ? ((equity - prevEq) / prevEq) * 100 : 0
    monthlyReturns.push({ month, return: Math.round(ret * 100) / 100, equity: Math.round(equity * 100) / 100 })
    prevEq = equity
  }

  return {
    strategy_name: s.name,
    start_date: bars[0].datetime.split("T")[0],
    end_date: bars[bars.length - 1].datetime.split("T")[0],
    initial_capital: initialCapital,
    final_capital: Math.round(finalCapital * 100) / 100,
    total_return_pct: Math.round(totalReturn * 100) / 100,
    cagr: Math.round(cagr * 100) / 100,
    max_drawdown_pct: Math.round(maxDD * 100) / 100,
    max_drawdown_amount: Math.round(peakEquity * maxDD / 100 * 100) / 100,
    volatility: Math.round(volatility * 100) / 100,
    sharpe_ratio: Math.round(sharpe * 100) / 100,
    sortino_ratio: Math.round(sortino * 100) / 100,
    win_rate: Math.round(winRate * 100) / 100,
    average_win: Math.round(avgWin * 100) / 100,
    average_loss: Math.round(avgLoss * 100) / 100,
    profit_factor: Math.round(profitFactor * 100) / 100,
    total_trades: n,
    winning_trades: wins.length,
    losing_trades: losses.length,
    avg_holding_bars: Math.round(avgBars * 10) / 10,
    exposure_pct: Math.round(exposure * 100) / 100,
    total_brokerage: Math.round(totalBrokerage * 100) / 100,
    total_stt: Math.round(totalStt * 100) / 100,
    equity_curve: equityCurve,
    trades,
    monthly_returns: monthlyReturns,
  }
}

File: backend/app/core/engine/backtest.py
import uuid
from datetime import datetime

import pandas as pd

from ..indicators.technical import compute_indicator
from ..models.results import (
    BacktestResult,
    ClosedTrade,
    EquityPoint,
    OpenPosition,
    TradeOrder,
)
from ..models.strategy import (
    Condition,
    ConditionOperator,
    StrategyDefinition,
)


class BacktestEngine:
    def __init__(self, strategy: StrategyDefinition, data: pd.DataFrame):
        self.strategy = strategy
        self.data = data.copy()
        self.capital = strategy.initial_capital
        self.equity = strategy.initial_capital
        self.peak_equity = strategy.initial_capital
        self.positions: list[OpenPosition] = []
        self.closed_trades: list[ClosedTrade] = []
        self.equity_curve: list[EquityPoint] = []
        self.total_brokerage = 0.0
        self.total_stt = 0.0
        self._indicator_cache: dict[str, pd.Series | pd.DataFrame] = {}
        self._prepare_data()

    def _prepare_data(self):
        for rule in self.strategy.entry_rules:
            for cond in rule.conditions:
                self._ensure_indicator(cond.indicator, cond.period or 14)
        for rule in self.strategy.exit_rules:
            for cond in rule.conditions:
                self._ensure_indicator(cond.indicator, cond.period or 14)

    def _ensure_indicator(self, name: str, period: int):
        cache_key = f"{name}_{period}"
        if cache_key not in self._indicator_cache:
            self._indicator_cache[cache_key] = compute_indicator(
                name, self.data, {"period": period}
            )

    def _get_indicator_value(
        self, name: str, period: int, idx: int
    ) -> float:
        cache_key = f"{name}_{period}"
        if cache_key not in self._indicator_cache:
            self._ensure_indicator(name, period)
        series = self._indicator_cache[cache_key]
        if isinstance(series, pd.DataFrame):
            if name == "macd":
                return float(series["macd"].iloc[idx]) if idx < len(series) else 0.0
            return 0.0
        val = series.iloc[idx]
        return float(val) if pd.notna(val) else 0.0

    def _evaluate_condition(self, cond: Condition, idx: int) -> bool:
        actual_val = self._get_indicator_value(cond.indicator, cond.period or 14, idx)
        try:
            target_val = float(cond.value) if isinstance(cond.value, (int, float, str)) else 0.0
        except (ValueError, TypeError):
            target_val = 0.0

        if cond.operator == ConditionOperator.GT:
            return actual_val > target_val
        elif cond.operator == ConditionOperator.LT:
            return actual_val < target_val
        elif cond.operator == ConditionOperator.GTE:
            return actual_val >= target_val
        elif cond.operator == ConditionOperator.LTE:
            return actual_val <= target_val
        elif cond.operator == ConditionOperator.EQ:
            return abs(actual_val - target_val) < 1e-8
        elif cond.operator == ConditionOperator.NEQ:
            return abs(actual_val - target_val) >= 1e-8
        elif cond.operator == ConditionOperator.CROSS_ABOVE:
            if idx < 1:
                return False
            prev = self._get_indicator_value(cond.indicator, cond.period or 14, idx - 1)
            return prev <= target_val and actual_val > target_val
        elif cond.operator == ConditionOperator.CROSS_BELOW:
            if idx < 1:
                return False
            prev = self._get_indicator_value(cond.indicator, cond.period or 14, idx - 1)
            return prev >= target_val and actual_val < target_val
        return False

    def _evaluate_conditions(self, conditions: list[Condition], idx: int) -> bool:
        return all(self._evaluate_condition(c, idx) for c in conditions)

    def _calculate_charges(self, price: float, quantity: float, side: str) -> dict:
        tc = self.strategy.transaction_costs
        turnover = price * quantity
        brokerage = tc.brokerage_per_order
        stt = turnover * (tc.stt_percent / 100)
        exchange = turnover * (tc.exchange_charges / 100)
        sebi = turnover * (tc.sebi_charges / 100)
        gst = (brokerage + exchange) * (tc.gst_percent / 100)
        slippage = tc.slippage_ticks * tc.slippage_value * quantity
        return {
            "brokerage": brokerage,
            "stt": stt,
            "other_charges": exchange + sebi + gst,
            "slippage": slippage,
        }

    def _fill_order(
        self, timestamp: datetime, instrument: str, side: str,
        quantity: float, price: float, tag: str = "",
    ) -> TradeOrder:
        tc = self.strategy.transaction_costs
        slippage = tc.slippage_ticks * tc.slippage_value
        fill_price = price + slippage if side == "buy" else price - slippage
        charges = self._calculate_charges(fill_price, quantity, side)
        total_cost = charges["brokerage"] + charges["stt"] + charges["other_charges"]
        if side == "buy":
            self.capital -= fill_price * quantity + total_cost
        else:
            self.capital += fill_price * quantity - total_cost
        self.total_brokerage += charges["brokerage"]
        self.total_stt += charges["stt"]

        return TradeOrder(
            id=str(uuid.uuid4()),
            timestamp=timestamp,
            instrument=instrument,
            side=side,
            quantity=quantity,
            price=fill_price,
            slippage=charges["slippage"],
            brokerage=charges["brokerage"],
            stt=charges["stt"],
            other_charges=charges["other_charges"],
            tag=tag,
        )

    def _compute_position_pnl(self, pos: OpenPosition, current_price: float) -> float:
        entry_price = pos.entry_order.price
        qty = pos.entry_order.quantity
        if pos.entry_order.side == "buy":
            return (current_price - entry_price) * qty
        else:
            return (entry_price - current_price) * qty

    def _check_exit_conditions(self, pos: OpenPosition, idx: int) -> str | None:
        current_price = float(self.data["Close"].iloc[idx])
        rm = self.strategy.risk_management

        if rm.stop_loss_percent and pos.entry_order.side == "buy":
            sl_price = pos.entry_order.price * (1 - rm.stop_loss_percent / 100)
            if current_price <= sl_price:
                return "stop_loss"
        elif rm.stop_loss_percent and pos.entry_order.side == "sell":
            sl_price = pos.entry_order.price * (1 + rm.stop_loss_percent / 100)
            if current_price >= sl_price:
                return "stop_loss"

        if rm.take_profit_percent and pos.entry_order.side == "buy":
            tp_price = pos.entry_order.price * (1 + rm.take_profit_percent / 100)
            if current_price >= tp_price:
                return "take_profit"
        elif rm.take_profit_percent and pos.entry_order.side == "sell":
            tp_price = pos.entry_order.price * (1 - rm.take_profit_percent / 100)
            if current_price <= tp_price:
                return "take_profit"

        if rm.trailing_stop_percent:
            if pos.entry_order.side == "buy":
                trail_price = pos.highest_price * (1 - rm.trailing_stop_percent / 100)
                if current_price <= trail_price:
                    return "trailing_stop"
            else:
                trail_price = pos.highest_price * (1 + rm.trailing_stop_percent / 100)
                if current_price >= trail_price:
                    return "trailing_stop"

        for rule in self.strategy.exit_rules:
            if rule.max_bars_in_trade and pos.bars_held >= rule.max_bars_in_trade:
                return "max_bars"
            if rule.exit_at_end_of_day:
                ts = self.data.index[idx]
                if hasattr(ts, "hour") and ts.hour >= 15 and ts.minute >= 15:
                    return "eod"
            if rule.conditions and self._evaluate_conditions(rule.conditions, idx):
                return "signal"

        return None

    def _get_quantity(self) -> float:
        ps = self.strategy.position_sizing
        if ps.sizing_type.value == "fixed_lot":
            return float(ps.lot_size)
        elif ps.sizing_type.value == "fixed_capital":
            price = float(self.data["Close"].iloc[0])
            return max(1, int((ps.capital_per_trade or 100000) / price))
        elif ps.sizing_type.value == "percent_capital":
            price = float(self.data["Close"].iloc[0])
            alloc = self.equity * (ps.percent_capital or 10) / 100
            return max(1, int(alloc / price))
        return 1.0

    def run(self) -> BacktestResult:
        for idx in range(len(self.data)):
            ts = self.data.index[idx]
            current_price = float(self.data["Close"].iloc[idx])

            closed_indices = []
            for i, pos in enumerate(self.positions):
                if current_price > pos.highest_price:
                    pos.highest_price = current_price
                pos.bars_held += 1
                pos.current_price = current_price
                pos.unrealized_pnl = self._compute_position_pnl(pos, current_price)

                exit_reason = self._check_exit_conditions(pos, idx)
                if exit_reason:
                    exit_order = self._fill_order(
                        ts, pos.entry_order.instrument,
                        "sell" if pos.entry_order.side == "buy" else "buy",
                        pos.entry_order.quantity, current_price, exit_reason,
                    )
                    net_pnl = pos.unrealized_pnl - exit_order.brokerage - exit_order.stt - exit_order.other_charges
                    self.closed_trades.append(ClosedTrade(
                        entry_order=pos.entry_order,
                        exit_order=exit_order,
                        pnl=pos.unrealized_pnl,
                        net_pnl=net_pnl,
                        holding_bars=pos.bars_held,
                        exit_reason=exit_reason,
                    ))
                    self.equity += pos.unrealized_pnl
                    closed_indices.append(i)

            for i in reversed(closed_indices):
                self.positions.pop(i)

            rm = self.strategy.risk_management
            daily_loss = 0
            for trade in self.closed_trades:
                if hasattr(trade.exit_order.timestamp, "date") and \
                   hasattr(ts, "date") and \
                   trade.exit_order.timestamp.date() == ts.date():
                    if trade.net_pnl < 0:
                        daily_loss += abs(trade.net_pnl)

            if rm.max_loss_per_day and daily_loss >= rm.max_loss_per_day:
                self.equity_curve.append(EquityPoint(
                    timestamp=ts, equity=self.equity, drawdown=0.0
                ))
                continue

            if len(self.positions) < rm.max_positions:
                for rule in self.strategy.entry_rules:
                    if self._evaluate_conditions(rule.conditions, idx):
                        qty = self._get_quantity()
                        order = self._fill_order(
                            ts, rule.instrument, rule.side.value,
                            qty, current_price, "entry",
                        )
                        self.positions.append(OpenPosition(
                            entry_order=order,
                            current_price=current_price,
                            highest_price=current_price,
                        ))
                        break

            if self.equity > self.peak_equity:
                self.peak_equity = self.equity
            dd = (self.peak_equity - self.equity) / self.peak_equity if self.peak_equity > 0 else 0
            self.equity_curve.append(EquityPoint(
                timestamp=ts, equity=self.equity, drawdown=dd * 100
            ))

        for pos in list(self.positions):
            last_price = float(self.data["Close"].iloc[-1])
            exit_order = self._fill_order(
                self.data.index[-1], pos.entry_order.instrument,
                "sell" if pos.entry_order.side == "buy" else "buy",
                pos.entry_order.quantity, last_price, "end_of_backtest",
            )
            pnl = self._compute_position_pnl(pos, last_price)
            self.closed_trades.append(ClosedTrade(
                entry_order=pos.entry_order,
                exit_order=exit_order,
                pnl=pnl,
                net_pnl=pnl - exit_order.brokerage - exit_order.stt,
                holding_bars=pos.bars_held,
                exit_reason="end_of_backtest",
            ))
        self.positions.clear()

        return self._compute_results()

    def _compute_results(self) -> BacktestResult:
        s = self.strategy
        trades = self.closed_trades
        n = len(trades)
        wins = [t for t in trades if t.net_pnl > 0]
        losses = [t for t in trades if t.net_pnl <= 0]

        total_days = (self.data.index[-1] - self.data.index[0]).days
        years = max(total_days / 365.25, 0.01)

        final = self.equity
        total_return = ((final - s.initial_capital) / s.initial_capital) * 100
        cagr = ((final / s.initial_capital) ** (1 / years) - 1) * 100

        eq_series = [e.equity for e in self.equity_curve]
        dd_series = [e.drawdown for e in self.equity_curve]
        max_dd = max(dd_series) if dd_series else 0

        daily_returns = []
        for i in range(1, len(eq_series)):
            if eq_series[i - 1] > 0:
                daily_returns.append((eq_series[i] - eq_series[i - 1]) / eq_series[i - 1])
        import numpy as np
        daily_ret_arr = np.array(daily_returns) if daily_returns else np.array([0])
        volatility = float(np.std(daily_ret_arr) * np.sqrt(252) * 100)
        avg_ret = float(np.mean(daily_ret_arr))
        downside = daily_ret_arr[daily_ret_arr < 0]
        downside_vol = float(np.std(downside) * np.sqrt(252)) if len(downside) > 0 else 1e-10
        sharpe = (avg_ret * 252) / (float(np.std(daily_ret_arr) * np.sqrt(252)) if np.std(daily_ret_arr) > 0 else 1e-10)
        sortino = (avg_ret * 252) / downside_vol

        avg_win = sum(t.net_pnl for t in wins) / len(wins) if wins else 0
        avg_loss = sum(t.net_pnl for t in losses) / len(losses) if losses else 0
        gross_profit = sum(t.net_pnl for t in wins)
        gross_loss = abs(sum(t.net_pnl for t in losses))
        profit_factor = gross_profit / gross_loss if gross_loss > 0 else 0
        win_rate = (len(wins) / n * 100) if n > 0 else 0
        avg_bars = sum(t.holding_bars for t in trades) / n if n > 0 else 0

        trading_days = len(eq_series)
        days_in_position = sum(1 for t in trades for _ in range(t.holding_bars))
        exposure = (days_in_position / trading_days * 100) if trading_days > 0 else 0

        monthly = {}
        for e in self.equity_curve:
            key = e.timestamp.strftime("%Y-%m")
            monthly[key] = e.equity

        monthly_returns = []
        prev = s.initial_capital
        for month, eq in sorted(monthly.items()):
            ret = ((eq - prev) / prev * 100) if prev > 0 else 0
            monthly_returns.append({"month": month, "return": round(ret, 2), "equity": round(eq, 2)})
            prev = eq

        return BacktestResult(
            strategy_name=s.name,
            start_date=str(self.data.index[0].date()),
            end_date=str(self.data.index[-1].date()),
            initial_capital=s.initial_capital,
            final_capital=round(final, 2),
            total_return_pct=round(total_return, 2),
            cagr=round(cagr, 2),
            max_drawdown_pct=round(max_dd, 2),
            max_drawdown_amount=round(self.peak_equity * max_dd / 100, 2),
            volatility=round(volatility, 2),
            sharpe_ratio=round(sharpe, 2),
            sortino_ratio=round(sortino, 2),
            win_rate=round(win_rate, 2),
            average_win=round(avg_win, 2),
            average_loss=round(avg_loss, 2),
            profit_factor=round(profit_factor, 2),
            total_trades=n,
            winning_trades=len(wins),
            losing_trades=len(losses),
            avg_holding_bars=round(avg_bars, 1),
            exposure_pct=round(exposure, 2),
            total_brokerage=round(self.total_brokerage, 2),
            total_stt=round(self.total_stt, 2),
            equity_curve=self.equity_curve,
            trades=trades,
            monthly_returns=monthly_returns,
        )
